### PR TITLE
feat(datasource): implement get_dataframe_by_tag for drift metric reference data

### PR DESCRIPTION
**Pull request by SudipSinha**
## Summary

All drift metric endpoints call `data_source.get_dataframe_by_tag(model_id, reference_tag)` to retrieve reference data, but the method doesn't exist on `DataSource`. Every drift computation fails with `AttributeError` and HTTP 500.

## Root cause

The drift endpoints were written to call `get_dataframe_by_tag` (matching the Java service's `filterRowsByTagEquals` pattern), but the method was never implemented on the Python `DataSource` class.

## How it works

The method reads all stored data for a model, then filters rows by tag:

1. Read input data and metadata via `ModelData.data()`
2. Find the `tags` column in metadata (column index 3, contains `list[str]` per row)
3. Build a boolean mask: `tag in row_tags` for each row
4. Apply the mask to the input data array
5. Return a pandas DataFrame with input columns only

**Returns input columns only** because drift metrics iterate over `request.fit_columns` (input feature names) and access `reference_df[feature_name]`.

**Reads all rows** (no batch size) because reference data tagged e.g. "TRAINING" could be anywhere in the dataset, not just the last N rows.

**Handles numpy arrays from `column_names()`** — `ModelData.column_names()` returns numpy arrays, not Python lists. The implementation converts to `list()` before calling `.index("tags")` to avoid `AttributeError`. Test mocks use `np.array()` to match the real return type and catch this class of bug.

## Files changed

| File | Change |
|------|--------|
| `src/service/data/datasources/data_source.py` | Add `get_dataframe_by_tag` method |
| `tests/service/data/datasources/test_datasource.py` | 3 tests: matching rows, non-existent tag, missing data (mocks use numpy arrays) |

## Test plan

- [x] All 24 datasource tests pass (3 new)
- [x] Test mocks use `np.array()` for `column_names()` return values, matching real `ModelData` behavior
- [x] `ruff check` clean
- [x] `ruff format` clean
- [x] `pyrefly check` clean
- [x] Rebuild container image and verify drift metric computation succeeds with tagged reference data

🤖 Generated with [Claude Code](https://claude.ai/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added support for generating a tag-filtered dataframe of a model’s input/output data, showing only rows that match a selected tag.
* **Bug Fixes**
  * Tag filtering now works reliably with tag metadata provided as arrays or lists.
  * Returns an empty dataframe when tag metadata or underlying data is missing, and surfaces dataframe construction failures consistently.
* **Tests**
  * Added coverage for tag-based filtering, including edge cases and empty-result scenarios.

### CODE REVIEW

File: src/service/data/datasources/data_source.py [hunk 1]
```diff
@@ -171,6 +172,62 @@ async def get_organic_dataframe(
...
+            model_id: The model ID
+            tag: The tag value to filter by (e.g. "TRAINING")
+
+        Returns:
+            A pandas DataFrame containing only rows whose tags list includes the tag
+
+        """
+        try:
+            model_data = ModelData(model_id)
+            input_data, _, metadata = await model_data.data()
```
Comment (RobGeada) [inline src/service/data/datasources/data_source.py]: Is there no need to get output_data here?

**Review by RobGeada — CHANGES_REQUESTED** [review]

File: src/service/data/datasources/data_source.py (hunk 1 above)
Comment (SudipSinha) [reply src/service/data/datasources/data_source.py]: Good catch — fixed in f0535ec. `output_data` and `output_names` are now included in the returned DataFrame alongside input columns. Tests updated to verify output columns are present.
